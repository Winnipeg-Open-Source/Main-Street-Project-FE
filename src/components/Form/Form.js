import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { Button, Flex, Text } from 'pcln-design-system';
import ButtonGroup from 'components/ButtonGroup';

const StyledForm = styled(Flex)`
    height: 100%;
`;

const Content = styled(Flex)`
    flex: 1;
    height: 100%;
    
    & > :not(:last-child) {
        margin-bottom: 8px;
    }
`;

function Form ({ className, title, children, onSaveClick, onCancelClick }) {
    return (
        <StyledForm className={className} flexDirection='column'>
            {title && <Text bold>{title}</Text>}
            <Content flexDirection='column' alignItems='flex-start' my={title ? 2 : 0}>
                {children}
            </Content>
            <ButtonGroup>
                {onCancelClick && <Button color='tertiary' variation='outline' width={1} onClick={onCancelClick}>Cancel</Button>}
                {onSaveClick && <Button width={1} onClick={onSaveClick}>Save</Button>}
            </ButtonGroup>
        </StyledForm>
    );
}

Form.displayName = 'Form';

Form.propTypes = {
    className: PropTypes.string,
    title: PropTypes.string,
    children: PropTypes.node,
    onSaveClick: PropTypes.func,
    onCancelClick: PropTypes.func,
};

Form.defaultProps = {
    className: '',
};

export default Form;
