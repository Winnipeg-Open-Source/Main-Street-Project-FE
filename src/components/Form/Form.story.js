import React from 'react';
import { Input } from 'pcln-design-system';
import Form from './Form';

const FormChildren = () => (
    <>
        <Input placeholder='First Input' />
        <Input placeholder='Second Input' />
        <Input placeholder='Third Input' />
    </>
);

export const WithoutTitle = () => (
    <Form>
        <FormChildren />
    </Form>
);

export const WithTitle = () => (
    <Form title='New Form'>
        <FormChildren />
    </Form>
);

export default {
    title: 'Form',
    component: Form,
};
