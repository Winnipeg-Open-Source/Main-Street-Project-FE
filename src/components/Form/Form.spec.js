import React from 'react';
import { fireEvent } from '@testing-library/react';
import { Input } from 'pcln-design-system';
import renderWithTheme from 'tests/helpers/renderWithTheme';
import Form from './Form';

const FormChildren = () => (
    <>
        <Input placeholder='First Input' />
        <Input placeholder='Second Input' />
        <Input placeholder='Third Input' />
    </>
);

describe ('Form', () => {
    it ('renders without crashing', () => {
        const { asFragment } = renderWithTheme(
            <Form>
                <FormChildren />
            </Form>
        );

        expect(asFragment()).toMatchSnapshot();
    });

    it ('renders without a title', () => {
        const { queryByText } = renderWithTheme(
            <Form>
                <FormChildren />
            </Form>
        );

        expect(queryByText('Title')).toBeNull();
    });

    it ('renders with a title', () => {
        const { queryByText } = renderWithTheme(
            <Form title='Title'>
                <FormChildren />
            </Form>
        );

        expect(queryByText('Title')).toBeDefined();
    });

    it ('renders with buttons', () => {
        const mockSaveClick = jest.fn();
        const mockCancelClick = jest.fn();

        const { getByText } = renderWithTheme(
            <Form title='Title' onCancelClick={mockCancelClick} onSaveClick={mockSaveClick}>
                <FormChildren />
            </Form>
        );

        const saveButton = getByText(/Save/);
        const cancelButton = getByText(/Cancel/);

        expect(mockSaveClick).toHaveBeenCalledTimes(0);
        fireEvent.click(saveButton);
        expect(mockSaveClick).toHaveBeenCalledTimes(1);

        expect(mockCancelClick).toHaveBeenCalledTimes(0);
        fireEvent.click(cancelButton);
        expect(mockCancelClick).toHaveBeenCalledTimes(1);
    });
});
