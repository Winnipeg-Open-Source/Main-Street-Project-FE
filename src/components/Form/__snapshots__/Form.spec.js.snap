// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form renders without crashing 1`] = `
<DocumentFragment>
  <div
    class="sc-pbJYR hNzbLI sc-pkfPC fQMEfn"
  >
    <div
      class="sc-AxjAm sc-AxiKw sc-qQxFs dNNRWZ"
    >
      <div
        class="sc-AxjAm sc-AxiKw sc-qYUWV sa-don"
      >
        <input
          class="sc-pTWqp fyIUKN"
          font-size="2,,1"
          placeholder="First Input"
        />
        <input
          class="sc-pTWqp fyIUKN"
          font-size="2,,1"
          placeholder="Second Input"
        />
        <input
          class="sc-pTWqp fyIUKN"
          font-size="2,,1"
          placeholder="Third Input"
        />
      </div>
      <div
        class="sc-AxjAm sc-AxiKw sc-psDhf iCfenm"
      />
    </div>
  </div>
</DocumentFragment>
`;
